src/prerender.ts

```diff
@@ -0,0 +1,303 @@
+	const type = element.type;
+	let s = `<${type}`;
+	const props = element.props || {};
+	let children = element.children;
+	for (const prop of Object.keys(props)) {
+		const value = props[prop];
+		// Filter out empty values:
+		if (value == null) continue;
+		if (prop === "children" || prop === "textContent") children = value;
+		else s += ` ${prop}="${enc(value)}"`;
+	}
+	s += ">";
+	if (!/link|meta|base/.test(type)) {
+		if (children) s += serializeElement(children);
+		s += `</${type}>`;
+	}
+	return s;
+}
+
+interface PrerenderPluginOptions {
+	prerenderScript?: string;
+	additionalPrerenderRoutes?: string[];
+}
+
+export function PrerenderPlugin({
+	prerenderScript,
+	additionalPrerenderRoutes,
+}: PrerenderPluginOptions = {}): Plugin {
+	const preloadHelperId = "vite/preload-helper";
+	let viteConfig: UserConfig = {};
+
+	additionalPrerenderRoutes ||= [];
+
+	/**
+	 * Retrieves the last, non-external script from the entry HTML document to use as
+	 * an additional input for Rollup. Hopefully this contains the user's `prerender()`
+	 * function.
+	 */
+	const guessPrerenderScriptFromHTML = async (input: InputOption) => {
+		// prettier-ignore
+		const entryHtml =
+			typeof input === "string"
+				? input
+				: Array.isArray(input)
+					? input.find(i => /html$/.test(i))
+					: Object.values(input).find(i => /html$/.test(i));
+
+		if (!entryHtml) throw new Error("Unable to detect entry HTML file.");
+
+		const htmlDoc = parse(await fs.readFile(entryHtml, "utf-8"));
+		const scripts = htmlDoc
+			.getElementsByTagName("script")
+			.map(s => s.getAttribute("src"))
+			.filter(src => src && !/^https:/.test(src));
+
+		const entryScript = scripts.reverse()[0];
+
+		if (!entryScript) throw new Error("Unable to detect local entry script");
+
+		return path.join(process.cwd(), viteConfig.root ?? "", entryScript);
+	};
+
+	return {
+		name: "headless-prerender",
+		apply: "build",
+		enforce: "post",
+		config(config) {
+			viteConfig = config;
+		},
+		async options(opts) {
+			if (!opts.input) return;
+			if (!prerenderScript) {
+				prerenderScript = await guessPrerenderScriptFromHTML(opts.input);
+			}
+
+			// prettier-ignore
+			opts.input =
+				typeof opts.input === "string"
+					? [opts.input, prerenderScript]
+					: Array.isArray(opts.input)
+						? [...opts.input, prerenderScript]
+						: { ...opts.input, prerenderEntry: prerenderScript };
+			opts.preserveEntrySignatures = "allow-extension";
+		},
+		// Injects a window check into Vite's preload helper, instantly resolving
+		// the module rather than attempting to add a <link> to the document.
+		transform(code, id) {
+			// Vite keeps changing up the ID, best we can do for cross-version
+			// compat is an `includes`
+			if (id.includes(preloadHelperId)) {
+				const s = new MagicString(code);
+				s.replace(
+					`deps.length === 0`,
+					`deps.length === 0 || typeof window === 'undefined'`,
+				);
```

Comment:
This bails out of the preloader which is all that's needed to get the default Vite setup working in Node: https://github.com/vitejs/vite/blob/b93dfe3e08f56cafe2e549efd80285a12a3dc2f0/packages/vite/src/node/plugins/importAnalysisBuild.ts#L95-L98

This has been stable from v2-v5 (didn't check 1) so I think it's pretty safe.

Reply:
Unfortunately they just inverted this in v5.0.5.

Open to suggestions if anyone has a better idea for a stable patch.